src/f32-gemm/MrxLmulv-rvv.c.in

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 SiFive, Inc.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert DATATYPE in ["F32"]
+$assert ACTIVATION in ["LINEAR", "RELU", "MINMAX"]
+$assert MR >= 1
+$assert LMUL in [1, 2, 4]
+#include <assert.h>
+
+#include <riscv_vector.h>
+
+#include <xnnpack/gemm.h>
+
+
+$DATATYPE_SPEC = {"F32": "f32"}[DATATYPE]
+$SUFFIX = {"LINEAR": "", "RELU": "_relu", "MINMAX": "_minmax"}[ACTIVATION]
+$PARAMS = {"LINEAR": "xnn_f32_default_params", "RELU": "xnn_f32_relu_params", "MINMAX": "xnn_f32_minmax_params"}[ACTIVATION]
+void xnn_${DATATYPE_SPEC}_gemm${SUFFIX}_ukernel_${MR}x${LMUL}v__rvv(
+    size_t mr,
+    size_t nc,
+    size_t kc,
+    const float* restrict a,
+    size_t a_stride,
+    const float* restrict w,
+    float* restrict c,
+    size_t cm_stride,
+    size_t cn_stride,
+    const union ${PARAMS} params[restrict XNN_MIN_ELEMENTS(1)])
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(kc % sizeof(float) == 0);
+  assert(a != NULL);
+  assert(w != NULL);
+  assert(c != NULL);
+
+  $if ACTIVATION == "MINMAX":
+    const float vmin = params->scalar.min;
+    const float vmax = params->scalar.max;
+  const float* a0 = a;
+  float* c0 = c;
+  $if MR > 1:
+    const float* a1 = (const float*) ((uintptr_t) a + a_stride);
+    float* c1 = (float*) ((uintptr_t) c + cm_stride);
+    $for M in range(2, MR):
+      const float* a${M} = (const float*) ((uintptr_t) a${M-1} + a_stride);
+      float* c${M} = (float*) ((uintptr_t) c${M-1} + cm_stride);
+
+  const size_t nr = __riscv_vsetvlmax_e32m${LMUL}();
+  size_t vl = nr;
+  if XNN_LIKELY(mr == ${MR}) {
+    do {
+      if XNN_UNLIKELY(nc < nr) {
+        vl = __riscv_vsetvl_e32m${LMUL}(nc);
+      }
+      nc = nc - vl;
+
+      vfloat32m${LMUL}_t v_acc0 =  __riscv_vle32_v_f32m${LMUL}(w, vl);
+      w = w + nr;
+      $for M in range(1, MR):  
+        vfloat32m${LMUL}_t v_acc${M} =  __riscv_vmv_v_v_f32m${LMUL}(v_acc0, vl);
+
+      size_t k = kc;
+      do {
+        vfloat32m${LMUL}_t v_w_k = __riscv_vle32_v_f32m${LMUL}(w, vl);
+        w = w + nr;
+        $for M in range(0, MR):  
+          v_acc${M} = __riscv_vfmacc_vf_f32m${LMUL}(v_acc${M}, *a${M}, v_w_k, vl);
```

Comment:
load the inputs first.  On arm it is
      $for M in range(MR):
        const float32x2_t va${M} = vld1_f32(a${M}); a${M} += 2;
 Each line should generate 1 instruction.  On ARM we expect the post increment to be part of the load instruction so its on the same line.
 On intel, we combine the load with a math instruction when we expect fusion.  Fro example, this is intel
 $for M in range(MR):
  const __m256i va${M}x01234567 = _mm256_xor_epi64(_mm256_set1_epi64x((int64_t) unaligned_load_u64(a${M})), vsign_mask);
  a${M} += 16;
generating a single XOR instruction with an embedded broadcast memory argument.  but the pointer increment is a separate add instruction

Reply:
k=1 is selected here to use the grouped vector registers with LMUL=4 fully.
LMUL=4 -> 8 grouped vector registers
You can see that 7 regs are used as accumulators. 1 is to load weight.